File: FHN.py
import numpy as np
import matplotlib.pyplot as plt
import sympy as sp

# parameters
a, b, c = 0.1, 0.01, 0.1

# grid for the phase plane
V_vals = np.linspace(-0.5, 1.5, 20)
w_vals = np.linspace(-0.2, 0.3, 20)
V, W = np.meshgrid(V_vals, w_vals)

# vector field
dV = V * (a - V) * (V - 1) - W
dW = b * V - c * W

# nullclines
V_range = np.linspace(-0.5, 1.5, 100)
W_nullcline = V_range * (a - V_range) * (V_range - 1)
V_nullcline = (b / c) * V_range

# equilibrium points
V_sym = sp.Symbol('V')
eq_expr = V_sym * (a - V_sym) * (V_sym - 1) - (b/c) * V_sym

V_solutions = sp.solve(eq_expr, V_sym)  # Solve for V
equilibria = [(float(V), float((b/c) * V)) for V in V_solutions]  # Compute (V*, w*)

# Plot the phase plane
plt.figure(figsize=(8, 6))
plt.streamplot(V, W, dV, dW, color="gray", linewidth=0.7, density=1.2)

# Plot nullclines
plt.plot(V_range, W_nullcline, 'r', label="V-nullcline")
plt.plot(V_range, V_nullcline, 'b', label="w-nullcline")

# Plot equilibrium points
for eq in equilibria:
    plt.plot(eq[0], eq[1], 'go', markersize=8)

plt.xlabel("Membrane Voltage (V)")
plt.ylabel("Recovery Variable (w)")
plt.title(f"Phase Plane of FitzHugh-Nagumo Model\n(a={a}, b={b}, c={c})")
plt.legend()
plt.grid()
plt.show()

# Print the equilibrium points
print("Equilibrium points:", equilibria)

File: FHN.m
%% FitzHugh-Nagumo Model Simulation using Euler and RK4 Methods

clc; clear; close all;

%% Parameters
a = 0.7; 
b = 0.8; 
c = 0.08;
I = 0.5;  % External current

t_range = [0 50]; % Time range (ms)
num_steps = 2000; 
dt = (t_range(2) - t_range(1)) / num_steps;
t = linspace(t_range(1), t_range(2), num_steps+1);

% Initial conditions
V0 = -1;
w0 = 0;

%% Euler's Method
V_euler = zeros(1, num_steps+1);
w_euler = zeros(1, num_steps+1);
V_euler(1) = V0;
w_euler(1) = w0;

for j = 1:num_steps
    dVdt = V_euler(j) * (a - V_euler(j)) * (V_euler(j) - 1) - w_euler(j) + I;
    dwdt = b * V_euler(j) - c * w_euler(j);
    
    V_euler(j+1) = V_euler(j) + dt * dVdt;
    w_euler(j+1) = w_euler(j) + dt * dwdt;
end

%% RK4 Method
V_rk4 = zeros(1, num_steps+1);
w_rk4 = zeros(1, num_steps+1);
V_rk4(1) = V0;
w_rk4(1) = w0;

for j = 1:num_steps
    k1_V = dt * (V_rk4(j) * (a - V_rk4(j)) * (V_rk4(j) - 1) - w_rk4(j) + I);
    k1_w = dt * (b * V_rk4(j) - c * w_rk4(j));
    
    k2_V = dt * ((V_rk4(j) + k1_V/2) * (a - (V_rk4(j) + k1_V/2)) * ((V_rk4(j) + k1_V/2) - 1) - (w_rk4(j) + k1_w/2) + I);
    k2_w = dt * (b * (V_rk4(j) + k1_V/2) - c * (w_rk4(j) + k1_w/2));
    
    k3_V = dt * ((V_rk4(j) + k2_V/2) * (a - (V_rk4(j) + k2_V/2)) * ((V_rk4(j) + k2_V/2) - 1) - (w_rk4(j) + k2_w/2) + I);
    k3_w = dt * (b * (V_rk4(j) + k2_V/2) - c * (w_rk4(j) + k2_w/2));
    
    k4_V = dt * ((V_rk4(j) + k3_V) * (a - (V_rk4(j) + k3_V)) * ((V_rk4(j) + k3_V) - 1) - (w_rk4(j) + k3_w) + I);
    k4_w = dt * (b * (V_rk4(j) + k3_V) - c * (w_rk4(j) + k3_w));
    
    V_rk4(j+1) = V_rk4(j) + (k1_V + 2*k2_V + 2*k3_V + k4_V) / 6;
    w_rk4(j+1) = w_rk4(j) + (k1_w + 2*k2_w + 2*k3_w + k4_w) / 6;
end

%% Plot Results (Separate Figures)

% Euler's Method Plots
figure;
plot(t, V_euler, 'r', 'LineWidth', 1.5);
xlabel('Time (ms)'); ylabel('Membrane Potential V');
title('Time vs Membrane Potential (Euler Method)');
grid on;

figure;
plot(t, w_euler, 'r', 'LineWidth', 1.5);
xlabel('Time (ms)'); ylabel('Recovery Variable w');
title('Time vs Recovery Variable (Euler Method)');
grid on;

figure;
plot(V_euler, w_euler, 'r', 'LineWidth', 1.5);
xlabel('Membrane Potential V'); ylabel('Recovery Variable w');
title('Phase Portrait (Euler Method)');
grid on;

% RK4 Method Plots
figure;
plot(t, V_rk4, 'b', 'LineWidth', 1.5);
xlabel('Time (ms)'); ylabel('Membrane Potential V');
title('Time vs Membrane Potential (RK4 Method)');
grid on;

figure;
plot(t, w_rk4, 'b', 'LineWidth', 1.5);
xlabel('Time (ms)'); ylabel('Recovery Variable w');
title('Time vs Recovery Variable (RK4 Method)');
grid on;

figure;
plot(V_rk4, w_rk4, 'b', 'LineWidth', 1.5);
xlabel('Membrane Potential V'); ylabel('Recovery Variable w');
title('Phase Portrait (RK4 Method)');
grid on;
